src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs

```diff
@@ -96,64 +99,95 @@ public ModelStateDictionary ModelState
         /// <summary>
         /// Gets or sets the <see cref="IUrlHelper"/>.
         /// </summary>
-        [Activate]
-        public IUrlHelper Url { get; set; }
+        public IUrlHelper Url
+        {
+            get
+            {
+                if (_url == null)
+                {
+                    // May be null in unit-testing scenarios.
+                    var services = ViewComponentContext?.ViewContext?.HttpContext?.RequestServices;
+                    _url = services?.GetRequiredService<IUrlHelper>();
+                }
 
-        /// <summary>
-        /// Gets or sets the <see cref="ViewContext"/>.
-        /// </summary>
-        [Activate]
-        public ViewContext ViewContext
+                return _url;
+            }
+
+            [param: NotNull]
+            set
+            {
+                _url = value;
+            }
+        }
+
+        [ViewComponentContext]
+        public ViewComponentContext ViewComponentContext
         {
             get
             {
-                if (_viewContext == null)
+                if (_viewComponentContext == null)
                 {
-                    // This should run only for the ViewComponent unit test scenarios
-                    _viewContext = new ViewContext();
+                    _viewComponentContext = new ViewComponentContext();
                 }
 
-                return _viewContext;
+                return _viewComponentContext;
             }
 
             [param: NotNull]
             set
             {
-                _viewContext = value;
+                _viewComponentContext = value;
             }
         }
 
         /// <summary>
-        /// Gets or sets the <see cref="ViewDataDictionary"/>.
+        /// Gets the <see cref="ViewContext"/>.
         /// </summary>
         [Activate]
+        public ViewContext ViewContext
+        {
+            get
+            {
+                return ViewComponentContext?.ViewContext;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the <see cref="ViewDataDictionary"/>.
+        /// </summary>
         public ViewDataDictionary ViewData
         {
             get
             {
-                if (_viewData == null)
+                return ViewComponentContext?.ViewData;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the <see cref="ICompositeViewEngine"/>.
+        /// </summary>
+        public ICompositeViewEngine ViewEngine
+        {
+            get
+            {
+                if (_viewEngine == null)
                 {
-                    // This should run only for the ViewComponent unit test scenarios
-                    _viewData = new ViewDataDictionary(ViewContext.ViewData);
+                    // May be null in unit-testing scenarios.
+                    var services = ViewComponentContext.ViewContext?.HttpContext?.RequestServices;
```

Comment:
ViewComponentContext is guaranteed to not be null. Removed some extra ?s from other places in code.